hal/stm32f7.c

```diff
@@ -0,0 +1,510 @@
+#define TIM2_APB1_CLOCK_ER_VAL     (1 << 0)
+#define PWR_APB1_CLOCK_ER_VAL   (1 << 28)
+
+#define APB2_CLOCK_ER (*(volatile uint32_t *)(0x40023844))
+#define APB2_CLOCK_RST          (*(volatile uint32_t *)(0x40023824))
+#define SYSCFG_APB2_CLOCK_ER (1 << 14)
+
+#define FLASH_BASE          (0x40023C00)
+#define FLASH_ACR           (*(volatile uint32_t *)(FLASH_BASE + 0x00))
+#define FLASH_KEYR          (*(volatile uint32_t *)(FLASH_BASE + 0x04))
+#define FLASH_OPTKEYR       (*(volatile uint32_t *)(FLASH_BASE + 0x08))
+#define FLASH_SR            (*(volatile uint32_t *)(FLASH_BASE + 0x0C))
+#define FLASH_CR            (*(volatile uint32_t *)(FLASH_BASE + 0x10))
+#define FLASH_OPTCR         (*(volatile uint32_t *)(FLASH_BASE + 0x14))
+
+/* Register values */
+#define FLASH_ACR_ARTRST                     (1 << 11)
+#define FLASH_ACR_PRFEN                       (1 << 9)
+#define FLASH_ACR_ARTEN                       (1 << 8)
+
+#define FLASH_SR_BSY                          (1 << 16)
+#define FLASH_SR_PGSERR                       (1 << 7)
+#define FLASH_SR_PGPERR                       (1 << 6)
+#define FLASH_SR_PGAERR                       (1 << 5)
+#define FLASH_SR_WRPERR                       (1 << 4)
+#define FLASH_SR_OPERR                        (1 << 1)
+#define FLASH_SR_EOP                          (1 << 0)
+
+#define FLASH_CR_LOCK                         (1 << 31)
+#define FLASH_CR_ERRIE                        (1 << 25)
+#define FLASH_CR_EOPIE                        (1 << 24)
+#define FLASH_CR_STRT                         (1 << 16)
+#define FLASH_CR_MER2                         (1 << 15)
+#define FLASH_CR_MER1                         (1 << 2)
+#define FLASH_CR_SER                          (1 << 1)
+#define FLASH_CR_PG                           (1 << 0)
+
+#define FLASH_CR_SNB_SHIFT                      3
+#define FLASH_CR_SNB_MASK                      0x1f
+
+#define FLASH_CR_PROGRAM_X8                   (0 << 8)
+#define FLASH_CR_PROGRAM_X16                  (1 << 8)
+#define FLASH_CR_PROGRAM_X32                  (2 << 8)
+#define FLASH_CR_PROGRAM_X64                  (3 << 8)
+
+#define FLASH_OPTCR_nDBOOT                     (1 << 28)
+#define FLASH_OPTCR_nDBANK                     (1 << 29)
+#define FLASH_OPTCR_STRT                       (1 << 1)
+
+#define FLASH_KEY1                            (0x45670123)
+#define FLASH_KEY2                            (0xCDEF89AB)
+
+#define FLASH_OPTKEY1                         (0x08192A3B)
+#define FLASH_OPTKEY2                         (0x4C5D6E7F)
+
+
+/* FLASH Geometry 
+ *
+ * per ST AN4826, two configurations are possible on STM32F7:
+ *
+ *  - Dual bank with swapping, 2 x 512KB banks, 8 sectors each
+ *  - Single bank, 1 x 1MB, 8 sectors in total
+ *
+ *  The chosen configuration depends on the FLASH_OPTCR_DBANK value.
+ *
+ *
+ * */
+
+#define FLASH_SECTOR_UNUSED    0xFFFFFFFF
+/*  This memory mapping is for 2MB flash (STM32F769). */
```

Comment:
When we use a device which has only 1MB flash, how should we configure? 
Can we use the same code?